packages/directive-functions-plugin/src/compilers.ts

```diff
@@ -460,15 +465,33 @@ export function findDirectives(
       `body.${topParentIndex}.declarations.0.init`,
     ) as SupportedFunctionPath
 
-    const [baseFilename, ..._searchParams] = opts.filename.split('?')
+    const [baseFilename, ..._searchParams] = opts.filename.split('?') as [
+      string,
+      ...Array<string>,
+    ]
     const searchParams = new URLSearchParams(_searchParams.join('&'))
     searchParams.set(opts.directiveSplitParam, '')
 
     const extractedFilename = `${baseFilename}?${searchParams.toString()}`
 
-    const functionId = makeFileLocationUrlSafe(
-      `${baseFilename}--${functionName}`.replace(opts.root, ''),
-    )
+    // Relative in order to have constant functionId regardless of the machine
+    // that we are executing
+    const relativeFilename = path.relative(opts.root, baseFilename)
+    let functionId = `${relativeFilename}--${functionName}`
+    if (opts.functionId) {
+      functionId = opts.functionId({ currentId: functionId })
+      // Handle cases in which the returned id conflicts with
+      // one of the already defined ids
+      if (functionId in directiveFnsById) {
+        let deduplicatedId = functionId
+        let iteration = 0
+        do {
+          deduplicatedId = `${deduplicatedId}_${++iteration}`
+        } while (deduplicatedId in directiveFnsById)
+        functionId = deduplicatedId
+      }
+    }
+    functionId = makeFunctionIdUrlSafe(functionId)
```

Comment:
i wouldnt postprocess the function id anymore.
the function id that is returned by the function id generator function must be url safe. we need to document this requirement.

then the invocation of `makeFunctionIdUrlSafe` can simply go into the default function id generator function for dev